runhouse/servers/http/http_server.py

```diff
@@ -127,34 +127,50 @@ def run_module(message: Message):
             fn_type,
             resources,
             conda_env,
+            run_name,
             args,
             kwargs,
         ] = b64_unpickle(message.data)
 
         try:
             args = obj_store.get_obj_refs_list(args)
             kwargs = obj_store.get_obj_refs_dict(kwargs)
+
             result = call_fn_by_type(
                 fn_type=fn_type,
                 fn_name=fn_name,
                 relative_path=relative_path,
                 module_name=module_name,
                 resources=resources,
                 conda_env=conda_env,
+                run_name=run_name,
                 args=args,
                 kwargs=kwargs,
                 serialize_res=True,
             )
             # We need to pin the run_key in the server's Python context rather than inside the call_fn context,
             # because the call_fn context is a separate process and the pinned object will be lost when Ray
             # garbage collects the call_fn process.
-            if fn_type == "remote":
-                (run_key, obj_ref) = result
+            from runhouse import Run
+
+            (res, obj_ref, run_key) = result
+
+            if obj_ref is not None:
                 obj_store.put_obj_ref(key=run_key, obj_ref=obj_ref)
-                result = pickle.dumps(run_key)
+
+            result = pickle.dumps(res) if isinstance(res, Run) else res
 
             HTTPServer.register_activity()
-            if isinstance(result, list):
+            if isinstance(result, ray.exceptions.RayTaskError):
```

Comment:
any other errors thrown would be caught and raised by the http server instead of returning the "result" object. For runs specifically we don't want to immediately raise an exception since we need to capture failed runs (and associated config, stderr, etc.) on the cluster's file system